src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -368,15 +364,14 @@ private static unsafe OperationStatus DecodeFromUtf8InPlace(Span<byte> buffer, o
                 uint sourceIndex = 0;
                 uint destIndex = 0;
 
-                // only decode input if it is a multiple of 4
-                if (bufferLength % 4 != 0)
+                if (TBase64Decoder.IsInValidLength(buffer.Length))
                 {
                     goto InvalidExit;
                 }
 
-                ref sbyte decodingMap = ref MemoryMarshal.GetReference(DecodingMap);
+                ref sbyte decodingMap = ref MemoryMarshal.GetReference(TBase64Decoder.DecodingMap);
 
-                while (sourceIndex < bufferLength - 4)
+                while (sourceIndex + 4 < bufferLength)
```

Comment:
Please revert this change. `sourceIndex` gets incremented in the loop, thus `sourceIndex + 4` needs to be re-computed. W/ `bufferLength - 4` that value can be kept in a register.
```suggestion
                while (sourceIndex < bufferLength - 4)
```

Reply:
Your comment makes sense, but for Base64Url `bufferLength` can be less than 4, in which case it overflows that is why I changed it, I will think about how to improve it

Reply:
Check `bufferLength` once before the loop, then these values can be kept in registers.

Reply:
It's actually not only the less than 4 case, I should have said for `Base64Url ` it overflows when `bufferLength` is not divisible by 4, so thinking to account the remainder in the `bufferLength - 4` part

Reply:
How is it overflowing when bufferLength is 5?  It doesn't update in this loop.

```C#
if (bufferLength >= 4)
{
    while (sourceIndex < bufferLength - 4)
    {
        ...
    }
}

switch (bufferLength - sourceIndex)
{
   ...
}
```

* sourceIndex: 0, bufferLength: 5
* Enter the if
* Enter the while
* Do work
* sourceIndex moves to 4, cycle the loop
* sourceIndex is greater than 1, so don't spin again
* exit the if
* Now process the 1 byte remaining.

Reply:
ah, right, forgot that `bufferLength` is not changing 😄